crates/rust-analyzer/src/caps.rs

```diff
@@ -69,29 +71,58 @@ pub fn server_capabilities(client_caps: &ClientCapabilities) -> ServerCapabiliti
         execute_command_provider: None,
         workspace: None,
         call_hierarchy_provider: Some(CallHierarchyServerCapability::Simple(true)),
-        semantic_tokens_provider: Some(
-            SemanticTokensOptions {
-                legend: SemanticTokensLegend {
-                    token_types: semantic_tokens::SUPPORTED_TYPES.to_vec(),
-                    token_modifiers: semantic_tokens::SUPPORTED_MODIFIERS.to_vec(),
-                },
+        semantic_tokens_provider,
+        experimental,
+    }
+}
 
-                document_provider: Some(SemanticTokensDocumentProvider::Bool(true)),
-                range_provider: Some(true),
-                work_done_progress_options: Default::default(),
+fn experimental_capabilities(client_caps: &ClientCapabilities) -> Option<Value> {
+    client_caps.experimental.as_ref().and_then(|it| {
```

Comment:
Do we actually need this? From my reading of the protocol, it's generally OK to add additional fields to experimental or on-experimental capabilities. Cap negotiations I think is only needed when the same of cap is changed? Like, when it used to be a boolean, but now an object.

Reply:
I think yes, we do. The documentation is not explicit about ServerCapabilities, but at least Visual Studio (not VSCode) fails immediately when it gets a server capability (experimental, semantic, etc.) not listed before in the ClientCapabilities. I think we shouldn't rely on vague documentation and just be as accurate as we can.

Reply:
This doesn't seem necessary to me. Unknown server capabilities should not crash the client. I think this is a bug in VS. It may be worth asking @dbaeumer

Reply:
Ok, then I'll leave only the dynamic_registration check for `textDocument/didSave`.

Reply:
The spec is very explict about client caps:

> For future compatibility a ClientCapabilities object literal can have more properties set than currently defined. Servers receiving a ClientCapabilities object literal with unknown properties should ignore these properties

It indeed is silent about server caps, but it's hard for me to imagine why it would be any different. And CodeAction kind specifically is designed around this strategy.

So I'd say that VS is buggy :)